packages/flutter/lib/src/material/button_bar.dart

```diff
@@ -172,3 +181,183 @@ class ButtonBar extends StatelessWidget {
+      ..mainAxisSize = mainAxisSize
+      ..crossAxisAlignment = crossAxisAlignment
+      ..textDirection = getEffectiveTextDirection(context)
+      ..verticalDirection = verticalDirection
+      ..textBaseline = textBaseline;
+  }
+}
+
+/// Attempts to display buttons in a row, but displays them in a column if
+/// there is not enough horizontal space.
+///
+/// It first attempts to lay out its buttons as though there were no
+/// maximumm width constraints on the widget. If the button bar's width is
+/// less than the maximum width constraints of the widget, it then lays
+/// out the widget as though it were placed in a [Row].
+///
+/// However, if the button bar's width exceeds the maximum width constraint on
+/// the widget, it then aligns its buttons in a column. The key difference here
+/// is that the [MainAxisAlignment] will then be treated as a
+/// cross-axis/horizontal alignment. For example, if the buttons overflow and
+/// [ButtonBar.alignment] was set to [MainAxisAligment.start], the buttons would
+/// align to the horizontal start of the button bar.
+class _RenderButtonBarRow extends RenderFlex {
+  /// Creates a button bar that attempts to display in a row, but displays in
+  /// a column if there is insufficient horizontal space.
+  _RenderButtonBarRow({
+    List<RenderBox> children,
+    Axis direction = Axis.horizontal,
+    MainAxisSize mainAxisSize = MainAxisSize.max,
+    MainAxisAlignment mainAxisAlignment = MainAxisAlignment.start,
+    CrossAxisAlignment crossAxisAlignment = CrossAxisAlignment.center,
+    TextDirection textDirection,
+    VerticalDirection verticalDirection = VerticalDirection.down,
+    TextBaseline textBaseline,
+  }) : super(
+    children: children,
+    direction: direction,
+    mainAxisSize: mainAxisSize,
+    mainAxisAlignment: mainAxisAlignment,
+    crossAxisAlignment: crossAxisAlignment,
+    textDirection: textDirection,
+    verticalDirection: verticalDirection,
+    textBaseline: textBaseline,
+  );
+
+  bool _hasCheckedLayoutWidth = false;
+
+  @override
+  BoxConstraints get constraints {
+    if (_hasCheckedLayoutWidth)
+      return super.constraints;
+    return super.constraints.copyWith(maxWidth: double.infinity);
+  }
+
+  @override
+  void performLayout() {
+    // Set check layout width to false in reload or update cases.
+    _hasCheckedLayoutWidth = false;
+
+    // Perform layout to ensure that button bar knows how wide it would
+    // ideally want to be.
+    super.performLayout();
+    _hasCheckedLayoutWidth = true;
+
+    // If the button bar is constrained by width and it overflows, set the
+    // buttons to align vertically. Otherwise, lay out the button bar
+    // horizontally.
+    if (size.width > constraints.maxWidth) {
+      RenderBox child = firstChild;
+      double currentHeight = 0.0;
+      while (child != null) {
+        final FlexParentData childParentData = child.parentData;
+        assert(child.parentData == childParentData);
```

Comment:
Presumably, this assert has already been applied by super.performLayout() or somewhere else in RenderFlex or its superclasses.

Reply:
I believe it has already. I'll remove it from here then